clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -512,12 +495,10 @@ private void completeBatch(ProducerBatch batch, ProduceResponse.PartitionRespons
                 final RuntimeException exception;
                 if (error == Errors.TOPIC_AUTHORIZATION_FAILED)
                     exception = new TopicAuthorizationException(batch.topicPartition.topic());
+                else if (error == Errors.CLUSTER_AUTHORIZATION_FAILED)
+                    exception = new ClusterAuthorizationException("The producer is not authorized to do idempotent sends");
```

Comment:
The problem is that the message for CLUSTER_AUTHORIZATION_FAILED is very generic since there are a number of cases that this handles. I wanted to be able to give a better message since we know the specific operation that would have failed.